src/DataStructures/LinkedMessageQueue.hpp

```diff
@@ -0,0 +1,178 @@
+    return std::hash<Id>{}(id.id);
+  }
+};
+}  // namespace std
+
+/// \cond
+template <typename Id, typename QueueTags>
+class LinkedMessageQueue;
+/// \endcond
+
+/// \ingroup DataStructuresGroup
+/// A collection of queues of asynchronously received data
+///
+/// This class is designed to collect asynchronously received
+/// messages, possibly from multiple sources, until sufficient data
+/// has been received for processing.  Each tag in \p QueueTags
+/// defines a queue of messages of type `Tag::type`.
+template <typename Id, typename... QueueTags>
+class LinkedMessageQueue<Id, tmpl::list<QueueTags...>> {
+  static_assert(sizeof...(QueueTags) > 0,
+                "Must have at least one message source.");
+
+ public:
+  using IdType = Id;
+
+  /// Insert data into a given queue at a given ID.  All queues must
+  /// receive data with the same collection of \p id_and_previous, but
+  /// are not required to receive them in the same order.
+  template <typename Tag>
+  void insert(const LinkedMessageId<Id>& id_and_previous,
+              typename Tag::type message) noexcept;
+
+  /// The next ID in the received sequence, if all queues have
+  /// received messages at that ID.
+  std::optional<Id> ready_id() const noexcept;
+
+  /// All the messages received at the time indicated by `ready_id()`.
+  /// It is an error to call this function if `ready_id()` would
+  /// return an empty optional.  This function removes the stored
+  /// messages from the queues, advancing the internal sequence ID.
+  tuples::TaggedTuple<QueueTags...> extract() noexcept;
+
+  // NOLINTNEXTLINE(google-runtime-references)
+  void pup(PUP::er& p) noexcept {
+    p | previous_id_;
+    p | messages_;
+  }
+
+ private:
+  template <typename Tag>
+  struct Optional {
+    using type = std::optional<typename Tag::type>;
+  };
+
+  using OptionalTuple = tuples::TaggedTuple<Optional<QueueTags>...>;
+
+  std::optional<Id> previous_id_{};
+  std::unordered_map<std::optional<Id>, std::pair<Id, OptionalTuple>>
+      messages_{};
+};
+
+template <typename Id, typename... QueueTags>
+template <typename Tag>
+void LinkedMessageQueue<Id, tmpl::list<QueueTags...>>::insert(
+    const LinkedMessageId<Id>& id_and_previous,
+    typename Tag::type message) noexcept {
+  static_assert((... or std::is_same_v<Tag, QueueTags>), "Unrecognized tag.");
+  const auto entry =
+      messages_
+          .insert({id_and_previous.previous,
+                   std::pair{id_and_previous.id, OptionalTuple{}}})
+          .first;
+  auto& value = entry->second;
+  auto& [id, tuple] = value;
+  ASSERT(id_and_previous.id == id,
+         "Received messages with different ids (" << id << " and "
+         << id_and_previous.id << ") but the same previous id ("
+         << id_and_previous.previous << ").");
+  ASSERT(not tuples::get<Optional<Tag>>(tuple).has_value(),
+         "Received duplicate messages at id " << id << " and previous "
+         << id_and_previous.previous << ".");
```

Comment:
If you can, include the tag in the ASSERT message so users know which tag already has a message at this id.

Reply:
The assertion message prints out the function, including specifying the template parameters.  Is that good enough?  Here's the current output, let me know if you would like a change:
```
Function: void LinkedMessageQueue<int, brigand::list<(anonymous namespace)::MyQueue<(anonymous namespace)::Label1>, (anonymous namespace)::MyQueue<(anonymous namespace)::Label2>>>::insert(const LinkedMessageId<Id> &, typename Tag::type) [Id = int, QueueTags = brigand::list<(anonymous namespace)::MyQueue<(anonymous namespace)::Label1>, (anonymous namespace)::MyQueue<(anonymous namespace)::Label2>>, Tag = (anonymous namespace)::MyQueue<(anonymous namespace)::Label1>]
Received duplicate messages at id 1 and previous id --.
```

Reply:
Oh yeah I see the `Tag = blah` at the end. That should enough.